dll/win32/kernel32/client/thread.c

```diff
@@ -1092,240 +1092,136 @@ QueueUserWorkItem(IN LPTHREAD_START_ROUTINE Function,
-        RtlClearBits(TlsBitmap, Index, 1);
-    }
-    else
-    {
-        /* Fail */
-        BaseSetLastNTError(STATUS_INVALID_PARAMETER);
-        RtlReleasePebLock();
-        return FALSE;
-    }
-
-    /* Done! */
+    
+    if (result != 0) 
+	{
+	NtSetInformationThread( GetCurrentThread(), ThreadZeroTlsCell, &index, sizeof(index) );
+	}
+    else 
+	{
+	SetLastError( ERROR_INVALID_PARAMETER );
     RtlReleasePebLock();
-    return TRUE;
+	}
+	
+    return result;
 }
 
-/*
- * @implemented
- */
-LPVOID
-WINAPI
-TlsGetValue(IN DWORD Index)
-{
-    PTEB Teb;
 
-    /* Get the TEB and clear the last error */
-    Teb = NtCurrentTeb();
-    Teb->LastErrorValue = 0;
+LPVOID WINAPI TlsGetValue( DWORD index )
+{
+    LPVOID result;
 
-    /* Check for simple TLS index */
-    if (Index < TLS_MINIMUM_AVAILABLE)
+    if (index < TLS_MINIMUM_AVAILABLE)
     {
-        /* Return it */
-        return Teb->TlsSlots[Index];
+        result = NtCurrentTeb()->TlsSlots[index];
     }
-
-    /* Check for valid index */
-    if (Index >= TLS_EXPANSION_SLOTS + TLS_MINIMUM_AVAILABLE)
+    if(index >= TLS_MINIMUM_AVAILABLE
     {
-        /* Fail */
-        BaseSetLastNTError(STATUS_INVALID_PARAMETER);
-        return NULL;
+        index = index - TLS_MINIMUM_AVAILABLE;
+		
+        if (index >= 8 * sizeof(NtCurrentTeb()->ProcessEnvironmentBlock->TlsExpansionBitmapBits))
+        {
+            SetLastError( ERROR_INVALID_PARAMETER );
+            result = NULL;
+        }
+		if (!NtCurrentTeb()->TlsExpansionSlots) 
+		{
+			result = NULL;
+		}
+        
+        else 
+		result = NtCurrentTeb()->TlsExpansionSlots[index];
     }
-
-    /* The expansion slots are allocated on demand, so check for it. */
-    Teb->LastErrorValue = 0;
-    if (!Teb->TlsExpansionSlots) return NULL;
-
-    /* Return the value from the expansion slots */
-    return Teb->TlsExpansionSlots[Index - TLS_MINIMUM_AVAILABLE];
+    SetLastError( ERROR_SUCCESS );
+    return result;
 }
 
-/*
- * @implemented
- */
-BOOL
-WINAPI
-TlsSetValue(IN DWORD Index,
-            IN LPVOID Value)
-{
-    DWORD TlsIndex;
-    PTEB Teb = NtCurrentTeb();
 
-    /* Check for simple TLS index */
-    if (Index < TLS_MINIMUM_AVAILABLE)
-    {
-        /* Return it */
-        Teb->TlsSlots[Index] = Value;
-        return TRUE;
-    }
 
-    /* Check if this is an expansion slot */
-    TlsIndex = Index - TLS_MINIMUM_AVAILABLE;
-    if (TlsIndex >= TLS_EXPANSION_SLOTS)
+BOOL WINAPI TlsSetValue( DWORD index, LPVOID value )
+{
+	bool result;
+    if (index < TLS_MINIMUM_AVAILABLE)
     {
-        /* Fail */
-        BaseSetLastNTError(STATUS_INVALID_PARAMETER);
-        return FALSE;
+        NtCurrentTeb()->TlsSlots[index] = value;
     }
-
-    /* Do we not have expansion slots? */
-    if (!Teb->TlsExpansionSlots)
+    if(index >= TLS_MINIMUM_AVAILABLE)
     {
-        /* Get the PEB lock to see if we still need them */
-        RtlAcquirePebLock();
-        if (!Teb->TlsExpansionSlots)
```

Comment:
Could you explain how does removing the PEB lock actually improve anything? If two or more concurrent threads are trying to allocate expansion slots from heap, would not that lead to erratic behaviors?